temporian/implementation/numpy_cc/operators/window.cc

```diff
@@ -10,436 +10,566 @@
+      if (curr_ts - v_timestamps[end_idx - 1] -
+              (curr_window_length - v_window_length[end_idx - 1]) >
+          0) {
+        // Window's beginning moved forward
+        while (begin_idx < n_event &&
+               v_timestamps[end_idx] - v_timestamps[begin_idx] >=
+                   curr_window_length) {
+          accumulator.Remove(v_values[begin_idx]);
+          begin_idx++;
+        }
+      } else {
+        // Window's beginning moved backwards.
+        // Note < instead of <= to respect (] window boundaries.
+        while (begin_idx > 0 &&
+               v_timestamps[end_idx] - v_timestamps[begin_idx - 1] <
+                   curr_window_length) {
+          begin_idx--;
+          accumulator.AddLeft(v_values[begin_idx]);
         }
       }
-      sum_values += value;
-      num_values++;
     }
 
-    void Remove(INPUT value) override
-    {
-      if constexpr (std::numeric_limits<INPUT>::has_quiet_NaN)
-      {
-        if (std::isnan(value))
-        {
-          return;
+    // Set current value of window to all values with the same timestamp.
+    const auto result = accumulator.Result();
+    for (size_t i = end_idx; i < first_diff_ts_idx; i++) {
+      v_output[i] = result;
+    }
+
+    // Move pointer to the index of the last value with the same timestamp.
+    end_idx = first_diff_ts_idx;
+  }
+
+  return output;
+}
+
+// External sampling, variable window length
+template <typename INPUT, typename OUTPUT, typename TAccumulator>
+py::array_t<OUTPUT> accumulate(const ArrayD &evset_timestamps,
+                               const py::array_t<INPUT> &evset_values,
+                               const ArrayD &sampling_timestamps,
+                               const ArrayD &window_length) {
+  // Input size
+  const size_t n_event = evset_timestamps.shape(0);
+  const size_t n_sampling = sampling_timestamps.shape(0);
+
+  // Allocate output array
+  auto output = py::array_t<OUTPUT>(n_sampling);
+
+  auto v_output = output.template mutable_unchecked<1>();
+  auto v_timestamps = evset_timestamps.unchecked<1>();
+  auto v_values = evset_values.template unchecked<1>();
+  auto v_sampling = sampling_timestamps.unchecked<1>();
+  auto v_window_length = window_length.unchecked<1>();
+
+  TAccumulator accumulator;
+
+  size_t begin_idx = 0;
+  size_t end_idx = 0;
+
+  for (size_t sampling_idx = 0; sampling_idx < n_sampling; sampling_idx++) {
+    const auto right_limit = v_sampling[sampling_idx];
+    const auto curr_window_length = v_window_length[sampling_idx];
+
+    while (end_idx < n_event && v_timestamps[end_idx] <= right_limit) {
+      accumulator.Add(v_values[end_idx]);
+      end_idx++;
+    }
+
+    if (end_idx > 0) {
+      // Move begin_idx forward or backwards depending on begin_diff.
+      if (right_limit - v_sampling[sampling_idx - 1] -
+              (curr_window_length - v_window_length[sampling_idx - 1]) >
+          0) {

```

Comment:
To improve readability, and factor code, what about creating a function:

bool contract_window(double ts, double last_ts, double window_length, double last_window_length)

returning (ts - last_ts - (window_length - last_window_length)) > 0
?